hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java

```diff
@@ -342,60 +365,94 @@ public HoodieInstant transitionCleanInflightToComplete(HoodieInstant inflightIns
+        createFileInMetaPath(fromInstant.getFileName(), data, false);
+        Path fromInstantPath = new Path(metaClient.getMetaPath(), fromInstant.getFileName());
+        Path toInstantPath = new Path(metaClient.getMetaPath(), toInstant.getFileName());
+        boolean success = metaClient.getFs().rename(fromInstantPath, toInstantPath);
+        if (!success) {
+          throw new HoodieIOException("Could not rename " + fromInstantPath + " to " + toInstantPath);
+        }
+      } else {
+        // Ensures old state exists in timeline
+        System.out.println("Checking for file exists ?" + new Path(metaClient.getMetaPath(),
+            fromInstant.getFileName()));
+        Preconditions.checkArgument(metaClient.getFs().exists(new Path(metaClient.getMetaPath(),
+            fromInstant.getFileName())));
+        // Use Write Once to create Target File
+        writeFileOnceInPath(new Path(metaClient.getMetaPath(), toInstant.getFileName()), data);
+        System.out.println("Create new file for toInstant ?" + new Path(metaClient.getMetaPath(), toInstant.getFileName()));
       }
     } catch (IOException e) {
       throw new HoodieIOException("Could not complete " + fromInstant, e);
     }
   }
 
-  private void revertStateTransition(HoodieInstant curr, HoodieInstant revert) {
-    Preconditions.checkArgument(curr.getTimestamp().equals(revert.getTimestamp()));
-    Path revertFilePath = new Path(metaClient.getMetaPath(), revert.getFileName());
+  private void revertCompleteToInflight(HoodieInstant completed, HoodieInstant inflight) {
+    Preconditions.checkArgument(completed.getTimestamp().equals(inflight.getTimestamp()));
+    Path inFlightCommitFilePath = new Path(metaClient.getMetaPath(), inflight.getFileName());
+    Path commitFilePath = new Path(metaClient.getMetaPath(), completed.getFileName());
     try {
-      if (!metaClient.getFs().exists(revertFilePath)) {
-        Path currFilePath = new Path(metaClient.getMetaPath(), curr.getFileName());
-        boolean success = metaClient.getFs().rename(currFilePath, revertFilePath);
-        if (!success) {
-          throw new HoodieIOException("Could not rename " + currFilePath + " to " + revertFilePath);
+      if (metaClient.getMetadataVersion().isNullVersion()) {
+        if (!metaClient.getFs().exists(inFlightCommitFilePath)) {
+          boolean success = metaClient.getFs().rename(commitFilePath, inFlightCommitFilePath);
+          if (!success) {
+            throw new HoodieIOException(
+                "Could not rename " + commitFilePath + " to " + inFlightCommitFilePath);
+          }
         }
-        log.info("Renamed " + currFilePath + " to " + revertFilePath);
+      } else {
+        Path requestedInstantFilePath = new Path(metaClient.getMetaPath(),
+            new HoodieInstant(State.REQUESTED, inflight.getAction(), inflight.getTimestamp()).getFileName());
+
+        // If inflight and requested files do not exist, create one
```

Comment:
what case will the inflight and requested not be present ?

Reply:
@n3nash : This can happen during the time of upgrade when the commit before we upgraded had to be reverted.